documentation/FIR-MLIR-dialect.md

```diff
@@ -0,0 +1,204 @@
+
+_Load_Expr_
+
+This operation promotes a reference to a Fortran object (for example, the result of a Locate_Expr operation) to an object value irrespective of type. It takes one argument, a reference to an abstract storage location.
+
+Recall that the abstract expression operations listed here operate on and produce SSA values. (Specifically, they are abstract operations with precise control- and data-flow constraints.)
+
+_Store_Expr_
+
+This operation demotes a value to a reference to an object. It takes two arguments: both a value and a reference to an abstract storage location (for example, the result of a Global_Expr operation) of the same type.
+
+Fortran has a number of mechanisms for specifying control flow, both structured (DO ... END DO) and unstructured (GOTO). Many of these can be directly lowered to the standard dialect's branch and conditional branch operations.
+
+There are a handful of multiway branch constructs to consider and these will be modeled as FIR (terminator) operations. The framework supports a generic terminator pattern operation. Specifically, a terminator is just an operation but it is augmented with successors (references to basic blocks) and successor arguments (for correct SSA form).
+
+_Select_
+
+Terminator operation for switching based on the return value of, for example, an I/O action.
+
+_Select_Case_
+
+Terminator operation for switching on the value of an expression.
+
+_Select_Rank_
+
+Terminator operation for switching on the rank attribute of an object. Must be lowered to the requisite operations on the object's descriptor (which shall assumably contain the rank, dimension, type, etc. of the Fortran object).
+
+_Select_Type_
+
+Terminator operation for switching on the type of an object. Must be lowered to the requisite operations on the object's type descriptor (which shall contain the encoded type of the Fortran object).
+
+_Unreachable_
+
+This operation is a terminator on a Block and indicates that control flow cannot reach this point. This happens when lowering Fortran's [ERROR] STOP statement into a runtime call. It is lowered into the LLVM IR as an unreachable instruction.
+
+For now, indirect branches (such as computed GOTO statements), will be lowered by mapping target blocks into an indirect index value used in a small chain of conditional branches.
+
+The standard dialect of MLIR supports calling procedures in full generality -- that is, both function calls in expressions and CALL statements to subroutines. Both cases will be lowered explicitly into FIR using the standard CallOp operation. The call operation is an abstract application of a (presumably) opaque set of computations on the SSA input values that produces a (set of) SSA result value(s). (Semantically, a CallOp is a named morphism like the anonymous ApplyExpr or LocateExpr.)
```

Comment:
standard "MLIR" callop operations?

Reply:
Standard is a dialect. I'll add "dialect" here.

Reply:
Note that will read as "standard dialect" in two consecutive sentences, which sounds rather stilted.

Reply:
Probably should define "standard" as a dialect somewhere then.

Reply:
I'll highlight where it was defined so it stands out.

Reply:
Done